### DISCUSSION TITLE
주어진 데이터와 모델에서 Cosine Similarity

### DISCUSSION THREAD

**Discussion by risolate** [boostcampaitech4nlp2/level1_semantictextsimilarity_nlp-level1-nlp-13#24] [General]
- [아이디어]
    Sentence Bert 에서 STS과제를 수행할 때, consine similarity를 사용한다고 한다. cosine similarity는 주어진 STS과제에 대해 얼마만큼의 상관성을 가지고 있을까?

- [실험방법]
    
    라벨값이 각각 0,1사이, 1,2사이, 2,3사이, 3,4사이, 4,5사이의 데이터를 모아 각각의 데이터셋과 데이터 로더를 만든다. 각각 데이터셋의 평균과 표준편차는 다음과 같다.

  | 평균 | 표준편차
-- | -- | --
label 0 to 1 | 0.2790278805822847 | 0.34744781289776006
label 1 to 2 | 1.5665610142630746 | 0.29493325741567195
label 2 to 3 | 2.643074324324324 | 0.2786624396091247
label 3 to 4 | 3.666683778234086 | 0.27444464838453153
label 4 to 5 | 4.411744583808439 | 0.2670424666234465

hugging face를 통해 klue/roberta-large를 불러온 직후 각각의 데이터들에 대해 어떠한 cosine similarity를 출력하는지 관찰해 보았다.

또한 비교를 위해 cosine similarity와 label 사이의 MSE loss를 학습한 모델로 동일한 출력을 확인하였다.
- [실험결과]
    
    hugging face를 통해 klue/roberta-large를 불러온 직후의 cosine similarity를 측정한후 plot한 결과 다음과 같이 나왔다. x축은 label의 값이며(0,1) (1,2) (2,3) (3,4) (4,5) y축은 cosine similiarity의 값이다(-1,1)

![Untitled1](https://user-images.githubusercontent.com/99873921/198521913-3e10e113-59ac-45e5-bb3c-7277e7baa505.png)

  | cosine similarity 평균 | 표준편차
-- | -- | --
label 0 to 1 | 0.9511237526098769 | 0.03890584178231344
label 1 to 2 | 0.9587980629713901 | 0.03273257910780431
label 2 to 3 | 0.9608527490416089 | 0.03135846308014975
label 3 to 4 | 0.9777674155436012 | 0.0165773577847248
label 4 to 5 | 0.9838194673798103 | 0.013838882451950337

1. 사전학습된 모델은 우리 데이터셋에 대해 전반적으로 높은 cosine similarity를 출력하였으며, 대부분 0.9~1.0사이에 값이 몰려있다.
2. label이 커질수록 구간의 cosine similarity평균도 증가하고 있으며 상관관계가 존재한다고 볼 수 있다.

이어서 SBERT를 사용해서 학습된 모델을 불러온 후 vaild dataset으로 같은 방법으로 plot한 결과는 다음과 같다.

![Untitled2](https://user-images.githubusercontent.com/99873921/198522229-0f3f9416-17a6-4e89-a182-bb72bc9ba104.png)

1. Sbert로 학습한 결과 label에 따라 cosine similarity가 점점 증가하는 모습을 보인다.
2. 다만 편차가 굉장히 큰 것을 관찰할 수 있다.

**Comment by ymnseol** [comment]
대회 초반에 cosine similarity & label scaling으로 같이 이야기를 해봤었는데, 과연 0\~5 범위의 label과 -1\~1 범위의 cosine similarity를 어떻게 범위를 맞춰줄 것인지 고민했었습니다.
Cosine similarity가 0에 가까울수록 서로 관련이 없는 문장이라고 생각을 했고, label도 0에 가까울수록 서로 관련이 없는 문장을 나타내서, label을 0~1로 scaling하고자 했는데, cosine similarity를 이렇게 이용하는 게 사용하는 게 맞는지 의문이 들었습니다.
STS에서 label과 cosine similarity의 scale이 일치하지 않을 때, 어떻게 처리하는 게 좋을까요?

**Reply by risolate** [reply]
이부분을 해결하고자 저도 마지막에 linear이나 feed forward network 등을 사용해 보았는데 결과가 좋지 않았습니다. 저도 다른 방법이 궁금합니다

**Reply by intrandom5** [reply]
+ 이걸 보고 모델이 STS 학습을 통해 서로 연관성이 적은 문장의 cosine similarity를 늘리는 방식으로 학습을 하는 것이고, 그렇기 때문에 라벨점수가 낮은 문장쌍보다 라벨 점수가 높은 문장 쌍을 더 잘 맞춘다라고 생각을 했었는데 일리 있는 해석인지?
+ 라벨을 -1~1로 normalize했을 때보다 0~1 사이로 normalize했을 때 결과가 더 잘 나왔었는데 왜 그럴지...? (이거 관련 discussion이 있었던 것 같은데 어디갔지)
